py/redrock/utils.py

```diff
@@ -284,6 +284,51 @@ def distribute_partition(A, k):
     return dist
 
 
+def distribute_work_lopsided(nproc, ids, weights=None, capacities=None):
```

Comment:
@dmargala does `distribute_work_lopsided` behave effectively the same as `distribute_work` when the `capacities` argument is not provided? If this is the case, I don't see why `distribute_work` shouldn't be replaced by `distribute_work_lopsided` in order to simplify maintenance.

Reply:
In that case, they should behave effectively the same although the ordering of partitioned task ids is different. The implementation in `distribute_work` looks pretty fancy and I wasn't sure how to modify it to incorporate per-process `capacities`. It's probably more efficient than my lazy `O(n^2)` implementation in `distribute_work_lopsided`. 

```python
In [1]: from redrock.utils import distribute_work, distribute_work_lopsided

In [2]: nprocs, ntasks = 4, 16

In [3]: ids = range(ntasks)

In [4]: distribute_work(nprocs, ids)
Out[4]: [[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11], [12, 13, 14, 15]]

In [5]: distribute_work_lopsided(nprocs, ids)
Out[5]: [[0, 4, 8, 12], [1, 5, 9, 13], [2, 6, 10, 14], [3, 7, 11, 15]]
```

Adding in weights:

```python
In [6]: weights = dict(zip(ids, [1] * ntasks))

In [7]: weights[3] = 5

In [8]: distribute_work(nprocs, ids, weights)
Out[8]: [[0, 1, 2], [3, 4, 5, 6], [7, 8, 9, 10, 11], [12, 13, 14, 15]]

In [9]: distribute_work_lopsided(nprocs, ids, weights)
Out[9]: [[3], [0, 4, 7, 10, 13], [1, 5, 8, 11, 14], [2, 6, 9, 12, 15]]
```

I'm not sure what `distribute_work` is doing there. Total weight per process using `distribute_work` is `3, 8, 5, 3`. Total weight per process using `distribute_work_lopsided` is `5, 5, 5, 5`.

Perhaps that should be spun off into a separate issue?

(while looking into this, I noticed that the weighted task sort in `distribute_work_lopsided` should be reversed  and pushed an update).

Reply:
@dmargala  I'm trying to understand how important `distribute_work_lopsided` is in the speedups you have reported. 

Above you mentioned that the "zfind" step is not gpu-ifyed while "zscan" is, so the allocation of processes using `distribute_work_lopsided` needs to be done separately for "zfind" and "zscan". This implies that "zfind" tasks depend on all "zscan" tasks to complete before running, but that there are situations in which some processes perform work for the "zscan" step on CPU (e.g. capacity=1) and others on GPU processes (e.g. capacity=15). Was this the case in your profiling? Naively I would expect only GPU processes to be working on tasks that have been optimized for GPU would be near optimal, with the gain from using CPUs on those tasks at the same time to be negligible.

Reply:
That's helpful indeed. Comparing V3 to V4 at numtasks = 4 to 16, it looks like adding the lopsided make the performance slightly worse. But it seems like there is no scaling test in these plot that shows whether lopsided by itself helps post-GPU_V5. I think it's fine keeping separate functions as is for now and coming back to this after testing on nightly processing and/or we have a better idea of the actual algorithmic differences between the two functions. Thanks.